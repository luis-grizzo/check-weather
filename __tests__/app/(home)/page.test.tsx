import { render, screen, fireEvent, waitFor } from '@testing-library/react'

import Home from '@/app/(home)/page'

const pushMock = jest.fn()

jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      push: pushMock
    }
  }
}))

describe('Home', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot', async () => {
    Object.defineProperty(navigator, 'permissions', {
      value: {
        query: jest.fn().mockResolvedValue({ state: 'prompt' })
      },
      writable: true
    })

    const { container } = render(<Home />)

    await waitFor(() => {
      expect(container).toMatchSnapshot()
    })
  })

  it('should try auto redirect if the permission is granted or denied', async () => {
    Object.defineProperty(navigator, 'permissions', {
      value: {
        query: jest.fn().mockResolvedValue({ state: 'granted' })
      },
      writable: true
    })

    Object.defineProperty(navigator, 'language', {
      value: 'en-US',
      writable: true
    })

    Object.defineProperty(navigator, 'geolocation', {
      value: {
        getCurrentPosition: jest
          .fn()
          .mockImplementation((success: PositionCallback) => {
            success({
              timestamp: 0,
              coords: {
                accuracy: 0,
                altitude: null,
                altitudeAccuracy: null,
                heading: null,
                speed: null,
                latitude: 10.101,
                longitude: 10.101
              }
            })
          })
      },
      writable: true
    })

    render(<Home />)

    const mockResponse = encodeURIComponent('10.1,10.1,imperial')

    await waitFor(() => {
      expect(pushMock).toHaveBeenCalledWith(`/${mockResponse}`)
    })
  })

  it('should redirect to location page with the correct imperial params', () => {
    Object.defineProperty(navigator, 'language', {
      value: 'en-US',
      writable: true
    })

    Object.defineProperty(navigator, 'geolocation', {
      value: {
        getCurrentPosition: jest
          .fn()
          .mockImplementation((success: PositionCallback) => {
            success({
              timestamp: 0,
              coords: {
                accuracy: 0,
                altitude: null,
                altitudeAccuracy: null,
                heading: null,
                speed: null,
                latitude: 10.101,
                longitude: 10.101
              }
            })
          })
      },
      writable: true
    })

    render(<Home />)

    const eButton = screen.getByRole('button')
    fireEvent.click(eButton)

    const mockResponse = encodeURIComponent('10.1,10.1,imperial')

    expect(pushMock).toHaveBeenCalledWith(`/${mockResponse}`)
  })

  it('should redirect to location page with the correct metric params', () => {
    Object.defineProperty(navigator, 'language', {
      value: 'pt-BR',
      writable: true
    })

    Object.defineProperty(navigator, 'geolocation', {
      value: {
        getCurrentPosition: jest
          .fn()
          .mockImplementation((success: PositionCallback) => {
            success({
              timestamp: 0,
              coords: {
                accuracy: 0,
                altitude: null,
                altitudeAccuracy: null,
                heading: null,
                speed: null,
                latitude: 10.101,
                longitude: 10.101
              }
            })
          })
      },
      writable: true
    })

    render(<Home />)

    const eButton = screen.getByRole('button')
    fireEvent.click(eButton)

    const mockResponse = encodeURIComponent('10.1,10.1,metric')

    expect(pushMock).toHaveBeenCalledWith(`/${mockResponse}`)
  })

  it('should throw if geolocation is denied', () => {
    Object.defineProperty(navigator, 'geolocation', {
      value: {
        getCurrentPosition: jest
          .fn()
          .mockImplementation((_, error: PositionErrorCallback) => {
            error?.({
              code: 1,
              message: 'message',
              PERMISSION_DENIED: 1,
              POSITION_UNAVAILABLE: 2,
              TIMEOUT: 3
            })
          })
      },
      writable: true
    })

    expect(() => {
      render(<Home />)

      const eButton = screen.getByRole('button')
      fireEvent.click(eButton)
    }).toThrow()
  })

  it('should throw if geolocation is undefined', () => {
    Object.defineProperty(navigator, 'geolocation', {
      value: undefined,
      writable: true
    })

    Object.defineProperty(navigator, 'language', {
      value: 'en-US',
      writable: true
    })

    expect(() => {
      render(<Home />)

      const eButton = screen.getByRole('button')
      fireEvent.click(eButton)
    }).toThrow()
  })
})
