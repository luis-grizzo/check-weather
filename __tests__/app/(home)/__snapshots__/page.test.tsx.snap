// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<div>
  <main
    class="flex flex-col items-center justify-center gap-8 h-dvh container mx-auto p-8"
  >
    <div
      class="flex items-center gap-4"
    >
      <img
        alt="Clear day icon"
        class="w-8 xs:w-9 sm:w-10 md:w-12 lg:w-[60px] xl:w-[72px] aspect-square"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="40"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <h1
        class="text-2xl xs:text-3xl sm:text-4xl md:text-5xl lg:text-6xl xl:text-7xl"
      >
        Check Weather
      </h1>
    </div>
    <p
      class="text-base text-center"
    >
      Consult your local weather info!
    </p>
    <button
      class="flex items-center justify-center py-3 px-4 bg-neutral-100/60 border-1 border-neutral-600/10 rounded-lg hover:bg-neutral-200/60 transition-colors"
    >
      Let's start!
    </button>
    <img
      alt="Clear day icon"
      class="absolute top-[calc(76px_-_96px)] -left-24 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateY(-100px) rotate(-48deg) translateZ(0);"
      width="40"
    />
    <img
      alt="Weather mix icon"
      class="hidden md:block absolute top-[calc(76px_-_96px)] left-1/2 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateY(-100px) translateX(-50%) translateZ(0);"
      width="40"
    />
    <img
      alt="Mist icon"
      class="absolute top-[calc(76px_-_96px)] -right-24 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateY(-100px) rotate(24deg) translateZ(0);"
      width="40"
    />
    <img
      alt="Airwave icon"
      class="hidden md:block absolute top-1/2 -left-24 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateX(-100px) translateY(-50%) translateZ(0);"
      width="40"
    />
    <img
      alt="Thunderstorm icon"
      class="hidden md:block absolute top-1/2 -right-24 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateX(100px) translateY(-50%) translateZ(0);"
      width="40"
    />
    <img
      alt="Foggy icon"
      class="absolute bottom-[calc(76px_-_96px)] -left-24 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateY(100px) rotate(24deg) translateZ(0);"
      width="40"
    />
    <img
      alt="Storm icon"
      class="hidden md:block absolute bottom-[calc(76px_-_96px)] left-1/2 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateY(100px) translateX(-50%) translateZ(0);"
      width="40"
    />
    <img
      alt="Clear night icon"
      class="absolute bottom-[calc(76px_-_96px)] -right-24 w-52 aspect-square -z-50"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent; opacity: 0; transform: translateY(100px) rotate(-48deg) translateZ(0);"
      width="40"
    />
  </main>
</div>
`;
